### PR TITLE
u2f: extract registration/authentication logic into lib/auth/u2f 

### PR DESCRIPTION
**Pull request by awly**
This is just a refactoring without functional changes. Pull all the u2f
handling spread across multiple client and server packages into one
place.

Also clean up an obsolete vendored dependency, unrelated to this PR.

### CODE REVIEW

**Review by a-palchikov — APPROVED** [review]

**Comment by webvictim** [comment]
This may be out of scope, but this doesn't work on MacOS 11.1:

```
~ » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=webvictim.teleport.sh:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: webvictim.teleport.sh:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=webvictim.teleport.sh:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: webvictim.teleport.sh:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=webvictim.teleport.sh:443, insecure=false) client/weblogin.go:303

ERROR REPORT:
Original Error: *errors.errorString hid: privilege violation
Stack Trace:
	/Users/gus/go/src/github.com/gravitational/teleport/lib/auth/u2f/authenticate.go:94 github.com/gravitational/teleport/lib/auth/u2f.AuthenticateSignChallenge
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/weblogin.go:514 github.com/gravitational/teleport/lib/client.SSHAgentU2FLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2314 github.com/gravitational/teleport/lib/client.(*TeleportClient).u2fLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2203 github.com/gravitational/teleport/lib/client.(*TeleportClient).localLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:1872 github.com/gravitational/teleport/lib/client.(*TeleportClient).Login
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:608 main.onLogin
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:440 main.Run
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:221 main.main
	/usr/local/go/src/runtime/proc.go:225 runtime.main
	/usr/local/go/src/runtime/asm_arm64.s:1130 runtime.goexit
User Message: hid: privilege violation

~ » ~/go/src/github.com/gravitational/teleport/build/tsh version
Teleport v6.0.0-alpha.1 git:v6.0.0-alpha.1-25-g44e0ac3d5 go1.16beta1

~ » go version
go version go1.16beta1 darwin/arm64
```

This may just be a limitation of what you can/can't do in a modern MacOS terminal. The first time I tried to run I got a different error, where my terminal had to request to receive keystrokes from any application. After I granted that permission and restarted the application, I got this error. I think it probably warrants investigation though as we have a lot of MacOS `tsh` users, some of whom may well be using U2F.

**Comment by webvictim** [comment]
With this said, I also get an error trying to use the 'old' `u2f-host` method as well:

```
~ » tsh -d login --proxy=webvictim.teleport.sh:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:801
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2282
DEBU [CLIENT]    not using loopback pool for remote proxy addr: webvictim.teleport.sh:443 client/api.go:2243
DEBU [CLIENT]    HTTPS client init(proxyAddr=webvictim.teleport.sh:443, insecure=false) client/weblogin.go:307
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: webvictim.teleport.sh:443 client/api.go:2243
DEBU [CLIENT]    HTTPS client init(proxyAddr=webvictim.teleport.sh:443, insecure=false) client/weblogin.go:307
Please press the button on your U2F key

ERROR REPORT:
Original Error: *trace.AccessDeniedError invalid U2F response
Stack Trace:
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/weblogin.go:584 github.com/gravitational/teleport/lib/client.SSHAgentU2FLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2202 github.com/gravitational/teleport/lib/client.(*TeleportClient).u2fLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2085 github.com/gravitational/teleport/lib/client.(*TeleportClient).localLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:1761 github.com/gravitational/teleport/lib/client.(*TeleportClient).Login
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:576 main.onLogin
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:431 main.Run
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:212 main.main
	/usr/local/go/src/runtime/proc.go:234 runtime.main
	/usr/local/go/src/runtime/asm_arm64.s:1131 runtime.goexit
User Message:

~ » file $(which u2f-host)
/opt/homebrew/bin/u2f-host: Mach-O 64-bit executable arm64
```

**Review by r0mant — COMMENTED** [review]
Haven't looked at the code yet, but noticed that github.com/flynn/u2f has a weird [license](https://github.com/flynn/u2f/blob/master/LICENSE). Not all licenses are compatible with Apache 2 as far as I know (GPL being a notorious example) and I had to pick a different library for MySQL integration than the one I had originally intended to use for that reason.

I would check whether their license is compatible with ours. Also, FWIW the other library you mentioned (https://github.com/marshallbrekka/go-u2fhost) is released under MIT which is ok.

File: lib/auth/u2f/authenticate.go
Comment (r0mant) [inline lib/auth/u2f/authenticate.go:1]: Missing license header.

File: lib/auth/u2f/authenticate.go
Comment (r0mant) [inline lib/auth/u2f/authenticate.go:143]: Is `tok.Authenticate()` blocking, or does it return `ErrPresenceRequired` if user doesn't press anything? Otherwise looks like a busy loop unless I'm missing something.

File: lib/auth/u2f/register.go
Comment (r0mant) [inline lib/auth/u2f/register.go:1]: Missing license header.

**Review by r0mant — COMMENTED** [review]
Don't have many comments besides the question about the library license compatibility.

**Review by awly — COMMENTED** [review]
@r0mant good point about the license.

re-implemented this on top of github.com/marshallbrekka/go-u2fhost - it's more complete, supports WebAuthn and maps much closer to our old github.com/tstranex/u2f (no data massaging needed).

the only downside is that on linux, it needs to be built with the `hidraw` build tag: https://github.com/marshallbrekka/go-u2fhost/tree/master#linux

@webvictim can you try the latest version of this PR on a mac?

File: lib/auth/u2f/authenticate.go
Comment (r0mant) [inline lib/auth/u2f/authenticate.go:167]: I would consider checking `res` for `nil` upon the exit from the outer loop, just in case.

**Review by r0mant — APPROVED** [review]

**Comment by webvictim** [comment]
@awly This worked (eventually) but I had many issues along the way. Maybe @r0mant or someone else could try this on an Intel Mac just to confirm that it isn't my Apple Silicon/beta version of Go causing problems?

`go version go1.16beta1 darwin/arm64`

Firstly, I logged out, then tried to log in and was told that the ID of the U2F device didn't match:

```
~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh logout
Logged out all users from all proxies.

~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=teleport.example.com:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303

ERROR REPORT:
Original Error: *trace.NotFoundError found 1 u2f devices, but none of them are registered with this Teleport user
Stack Trace:
	/Users/gus/go/src/github.com/gravitational/teleport/lib/auth/u2f/authenticate.go:137 github.com/gravitational/teleport/lib/auth/u2f.AuthenticateSignChallenge
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/weblogin.go:514 github.com/gravitational/teleport/lib/client.SSHAgentU2FLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2314 github.com/gravitational/teleport/lib/client.(*TeleportClient).u2fLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2203 github.com/gravitational/teleport/lib/client.(*TeleportClient).localLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:1872 github.com/gravitational/teleport/lib/client.(*TeleportClient).Login
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:608 main.onLogin
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:440 main.Run
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:221 main.main
	/usr/local/go/src/runtime/proc.go:225 runtime.main
	/usr/local/go/src/runtime/asm_arm64.s:1130 runtime.goexit
User Message: found 1 u2f devices, but none of them are registered with this Teleport user
```

This was the same U2F key and laptop that I used for the previous test, which had been registered using Chrome. I didn't update the server at any point during this test.

To get around this, I did a password reset and re-registered the key again, then tried to log in a second time. During this run, the Yubikey flashed a couple of times when the `Please press the button on your U2F key` message appeared, but it stopped flashing and generated the error shown before I could touch it (within ~1.5 seconds):

```
~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=teleport.example.com:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Please press the button on your U2F key

ERROR REPORT:
Original Error: *errors.errorString hidapi: hid_error is not implemented yet
Stack Trace:
	/Users/gus/go/src/github.com/gravitational/teleport/lib/auth/u2f/authenticate.go:161 github.com/gravitational/teleport/lib/auth/u2f.AuthenticateSignChallenge
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/weblogin.go:514 github.com/gravitational/teleport/lib/client.SSHAgentU2FLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2314 github.com/gravitational/teleport/lib/client.(*TeleportClient).u2fLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2203 github.com/gravitational/teleport/lib/client.(*TeleportClient).localLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:1872 github.com/gravitational/teleport/lib/client.(*TeleportClient).Login
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:608 main.onLogin
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:440 main.Run
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:221 main.main
	/usr/local/go/src/runtime/proc.go:225 runtime.main
	/usr/local/go/src/runtime/asm_arm64.s:1130 runtime.goexit
User Message: hidapi: hid_error is not implemented yet
```

At this point, I pressed up arrow and re-ran the command to see if it would happen again, but this time I didn't get a chance to touch the key before it logged me in (it went straight past the prompt, the Yubikey didn't even flash once):

```
~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=teleport.example.com:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Please press the button on your U2F key
DEBU [KEYAGENT]  Adding CA key for teleport.example.com client/keyagent.go:279
DEBU [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:N4EnjYU5XMgo3wsgdjXwou2HoINZyVnTpbzsHZb/V/g client/keystore.go:553
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYAGENT]  Adding CA key for teleport.example.com client/keyagent.go:279
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:N4EnjYU5XMgo3wsgdjXwou2HoINZyVnTpbzsHZb/V/g client/keystore.go:553
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [CLIENT]    Found clusters: [{"name":"teleport.example.com","lastconnected":"2021-01-25T14:08:16.506455292Z","status":"online"}] client/client.go:108
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:14 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:14 +0000 UTC". client/keystore.go:355
> Profile URL:        https://teleport.example.com:443
  Logged in as:       gus
  Cluster:            teleport.example.com
  Roles:              admin*
  Traits:             logins: []
  Logins:             root
  Kubernetes:         enabled
  Kubernetes cluster: "teleport.example.com"
  Valid until:        2021-01-25 22:08:14 -0400 AST [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty


* RBAC is only available in Teleport Enterprise
  https://goteleport.com/teleport/docs/enterprise
```

So I logged out, then tried logging in again:

```
~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh logout
Logged out all users from all proxies.

~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=teleport.example.com:443 --auth=local --user=gus
\INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303

ERROR REPORT:
Original Error: *trace.AccessDeniedError bad auth credentials
Stack Trace:
	/Users/gus/go/src/github.com/gravitational/teleport/lib/httplib/httplib.go:130 github.com/gravitational/teleport/lib/httplib.ConvertResponse
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/https_client.go:76 github.com/gravitational/teleport/lib/client.(*WebClient).PostJSON
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/weblogin.go:500 github.com/gravitational/teleport/lib/client.SSHAgentU2FLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2314 github.com/gravitational/teleport/lib/client.(*TeleportClient).u2fLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:2203 github.com/gravitational/teleport/lib/client.(*TeleportClient).localLogin
	/Users/gus/go/src/github.com/gravitational/teleport/lib/client/api.go:1872 github.com/gravitational/teleport/lib/client.(*TeleportClient).Login
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:608 main.onLogin
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:440 main.Run
	/Users/gus/go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:221 main.main
	/usr/local/go/src/runtime/proc.go:225 runtime.main
	/usr/local/go/src/runtime/asm_arm64.s:1130 runtime.goexit
User Message: bad auth credentials
```

I had no idea what was going on here, so logged out and tried logging in again. At this point, the Yubikey continued to flash at the `Please press the button on your U2F key` prompt and waited for me to touch it as expected. I did so, and it logged me in:

```
~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh logout
Logged out all users from all proxies.

~/go/src/github.com/gravitational/teleport(andrew/cli-native-u2f) » ~/go/src/github.com/gravitational/teleport/build/tsh -d login --proxy=teleport.example.com:443 --auth=local --user=gus
INFO [CLIENT]    no host login given. defaulting to gus client/api.go:892
INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.WI6JlKStse/Listeners" client/api.go:2393
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Enter password for Teleport user gus:
DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:2354
DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client/weblogin.go:303
Please press the button on your U2F key
DEBU [KEYAGENT]  Adding CA key for teleport.example.com client/keyagent.go:279
DEBU [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:N4EnjYU5XMgo3wsgdjXwou2HoINZyVnTpbzsHZb/V/g client/keystore.go:553
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYAGENT]  Adding CA key for teleport.example.com client/keyagent.go:279
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:N4EnjYU5XMgo3wsgdjXwou2HoINZyVnTpbzsHZb/V/g client/keystore.go:553
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [CLIENT]    Found clusters: [{"name":"teleport.example.com","lastconnected":"2021-01-25T14:08:38.673189267Z","status":"online"}] client/client.go:108
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
INFO [CLIENT]    Connecting proxy=teleport.example.com:3023 login='root' method=0 client/api.go:1787
DEBU [KEYAGENT]  Validated host teleport.example.com:3023. client/keyagent.go:321
INFO [CLIENT]    Successful auth with proxy teleport.example.com:3023 client/api.go:1797
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client/client.go:480
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
DEBU [KEYSTORE]  Returning SSH certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-cert.pub" valid until "2021-01-25 22:08:36 -0400 AST", TLS certificate "/Users/gus/.tsh/keys/teleport.example.com/gus-x509.pem" valid until "2021-01-26 02:08:36 +0000 UTC". client/keystore.go:355
> Profile URL:        https://teleport.example.com:443
  Logged in as:       gus
  Cluster:            teleport.example.com
  Roles:              admin*
  Traits:             logins: []
  Logins:             root
  Kubernetes:         enabled
  Kubernetes cluster: "teleport.example.com"
  Valid until:        2021-01-25 22:08:36 -0400 AST [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty


* RBAC is only available in Teleport Enterprise
  https://goteleport.com/teleport/docs/enterprise
```

**Comment by awly** [comment]
@webvictim a few ideas:

Your initial error
> User Message: found 1 u2f devices, but none of them are registered with this Teleport user

could also mean that the initial u2f token interaction failed. I was trying to avoid logging from a library, but in this case it's useful. Added some warning logs when talking to the u2f device fails.

> hidapi: hid_error is not implemented yet

comes from some low-level C code in the HID/libusb bindings from a cgo dependency that I don't understand :(

> I pressed up arrow and re-ran the command to see if it would happen again, but this time I didn't get a chance to touch the key before it logged me in (it went straight past the prompt, the Yubikey didn't even flash once)

That is surprising. Although maybe the u2f key "caches" user touch for a few seconds, since it verifies the touch via polling from the userspace.
If you manage to repro this reliably, let me know.

> User Message: bad auth credentials

This usually means a typo in the password, I think we validate it before sending the U2F challenge.

Sorry for the overall crappy experience, I'll work on ironing out all those problems.

**Comment by awly** [comment]
Still needs another approval here (probably for `docs/` and `lib/web/`)

**Comment by awly** [comment]
Ping @alex-kovoy, @fspmarshall, @kimlisa, @russjones, @webvictim, @benarent, and/or @klizhentas (because github won't tell me which files are missing approvals and I don't know what the minimum set of approvers should be)

**Review by klizhentas — CHANGES_REQUESTED** [review]
* `libusb` is released under the [GNU LGPL 2.1](https://github.com/libusb/libusb/blob/master/COPYING)license.


@awly this means we can't really use this library without digging more about how exactly are we integrating it. Are there any alternatives?

**Comment by webvictim** [comment]
> If you manage to repro this reliably, let me know.

I haven't managed to. Things worked fine when trying these latest changes a few times. My guess is that I did indeed typo the password at some point.

**Comment by awly** [comment]
@klizhentas on Linux we use `-tags hidraw` when compiling, which means using `hidraw` instead of `libusb`: https://github.com/bearsh/hid#libusb-backend-selection-linux-only

On other platforms, the underlying `hidapi` library ends up as BSD 3-clause license: https://github.com/libusb/hidapi/blob/master/LICENSE.txt

IIUC, this means we're not using `libusb`. I'm not an expert on licensing so please correct me if I'm wrong

**Comment by awly** [comment]
OK, while we sort out the licenses I removed both replacement u2f libraries.
This means we're still stuck with using `u2f-host` and this PR is just a refactoring for u2f logic.
PTAL

**Comment by awly** [comment]
Ping @klizhentas for approval

**Review by klizhentas — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**
